### DISCUSSION TITLE
Infinite loop in NextJS client component no matter what I do

### DISCUSSION THREAD

**Discussion by NixBiks** [statelyai/xstate#4877] [Q&A] [answered]
I've a component that looks like this

```typescript
"use client";

import { useMachine } from "@xstate/react";
import { setup } from "xstate";

function Component() {
  const [state, dispatch] = useMachine(
    setup({}).createMachine({
      initial: "idle",
      states: {
        idle: {},
      },
    })
  );

  return <div>Hello world!</div>;
}
```

If I try to render that in my NextJS application then I get 

    Error: Too many re-renders. React limits the number of renders to prevent an infinite loop.

What am I doing wrong here?

**Comment by Andarist** [comment, marked as answer]
You create new machines on each render. Try this:
```ts
"use client";

import { useMachine } from "@xstate/react";
import { setup } from "xstate";

const machine = setup({}).createMachine({
  initial: "idle",
  states: {
    idle: {},
  },
})

function Component() {
  const [state, dispatch] = useMachine(machine);

  return <div>Hello world!</div>;
}
```

**Reply by NixBiks** [reply]
Gosh of course. But how do you provide actions from other hooks then? Maybe I have a toast action from useToast that I'd like to call on certain events. 

Is [this example](https://stately.ai/blog/2022-07-18-just-use-hooks-xstate-in-react-components#a-custom-machine-hook) valid solution? Instead of giving a machine to useMachine then a function is given and won't cause rerender?

**Reply by Andarist** [reply]
You can supply render-dependent actions like this:
```ts
function Component() {
  const [state, dispatch] = useMachine(machine.provide({ actions: { doStuff: () => {} } }));

  return <div>Hello world!</div>;
}
```
